### PR TITLE
AuroraToEth Fast Bridge: Ether transfer

### PR DESCRIPTION
**Pull request by olga24912**
* supprot Ether transfers from Aurora
* make Aurora Fast Bridge contract shorter
* refactor integration tests

### CODE REVIEW

File: aurora/contracts/src/AuroraErc20FastBridge.sol [hunk 1]
```diff
@@ -320,36 +326,53 @@ contract AuroraErc20FastBridge is Initializable, UUPSUpgradeable, AccessControlU
         IEvmErc20 token = tokenInfo.auroraTokenAddress;
         require(tokenInfo.isStorageRegistered == true, "The token storage is not registered");
 
-        require(address(token) != address(0), "The token is not registered");
+        require(address(token) != address(0) ||
```
Comment (karim-en) [inline aurora/contracts/src/AuroraErc20FastBridge.sol:329]: let's add a variable `bool  isNativeTransfer = address(token) != address(0)`

File: aurora/contracts/src/AuroraErc20FastBridge.sol [hunk 2]
```diff
@@ -320,36 +326,53 @@ contract AuroraErc20FastBridge is Initializable, UUPSUpgradeable, AccessControlU
...
+            require(msg.value == 0, "Incorrect amount of Ether attached");
 
-        // WARNING: The `withdrawToNear` method works asynchronously.
-        // As a result, there is no guarantee that this method will be completed before `initTransfer()`.
-        // In case of such an error, the user will be able to call the `withdraw()` method and get his tokens back.
-        // We expect such an error not to happen as long as transactions are executed in one shard.
-        token.withdrawToNear(bytes(getImplicitNearAccountIdForSelf()), totalTokenAmount);
+            token.transferFrom(msg.sender, address(this), totalTokenAmount);
+
+            // WARNING: The `withdrawToNear` method works asynchronously.
```
Comment (karim-en) [inline aurora/contracts/src/AuroraErc20FastBridge.sol:340]: This comment can be removed due to the this explanation https://github.com/aurora-is-near/aurora-contracts-sdk/pull/13#discussion_r1330092099

File: aurora/contracts/src/AuroraErc20FastBridge.sol [hunk 3]
```diff
@@ -320,36 +326,53 @@ contract AuroraErc20FastBridge is Initializable, UUPSUpgradeable, AccessControlU
...
 
-        require(address(token) != address(0), "The token is not registered");
+        require(address(token) != address(0) ||
+                UtilsFastBridge.isStrEqual(transferMessage.transferTokenAccountIdOnNear, auroraEngineAccountIdOnNear),
+                "The token is not registered");
 
         uint256 totalTokenAmount = uint256(transferMessage.transferTokenAmount + transferMessage.feeTokenAmount);
 
-        token.transferFrom(msg.sender, address(this), totalTokenAmount);
+        if (address(token) != address(0)) {
```
Comment (karim-en) [inline aurora/contracts/src/AuroraErc20FastBridge.sol:335]: Let's use the `isNativeTransfer` here

File: aurora/contracts/src/AuroraErc20FastBridge.sol [hunk 4]
```diff
@@ -541,35 +559,44 @@ contract AuroraErc20FastBridge is Initializable, UUPSUpgradeable, AccessControlU
...
-                UtilsFastBridge.addressToString(recipient),
-                '"}'
-            )
-        );
         balance[token][recipient] -= recipientBalance;
 
+        string memory msgStr = UtilsFastBridge.addressToString(recipient);
+        if (UtilsFastBridge.isStrEqual(token, auroraEngineAccountIdOnNear)) {
+            string memory sixtyFourZeros;
+            for (uint8 i = 0; i < 64; i++) {
```
Comment (karim-en) [inline aurora/contracts/src/AuroraErc20FastBridge.sol:567]: This increase the gas consumption, let's try to use a const for it:
`string constant sixtyFourZeros = "0000000000000000000000000000000000000000000000000000000000000000";`, or try to find an more optimised solution if needed.

**Review by karim-en — CHANGES_REQUESTED** [review]

File: aurora/contracts/src/AuroraErc20FastBridge.sol [hunk 5]
```diff
@@ -186,6 +186,11 @@ contract AuroraErc20FastBridge is Initializable, UUPSUpgradeable, AccessControlU
       * - Call the get_erc20_from_nep141 function for extract the Aurora token address
     */
     function registerToken(string calldata nearTokenAccountId) external {
+        if (UtilsFastBridge.isStrEqual(nearTokenAccountId, auroraEngineAccountIdOnNear)) {
```
[the remark this answers is not in the document]
Comment (olga24912) [reply aurora/contracts/src/AuroraErc20FastBridge.sol]: fixed: https://github.com/aurora-is-near/fast-bridge-protocol/pull/113/commits/f8458df9770afc401b935a3e596d8da73a0520c3

File: aurora/contracts/src/AuroraErc20FastBridge.sol (hunk 1 above)
Comment (olga24912) [reply aurora/contracts/src/AuroraErc20FastBridge.sol]: fixed: https://github.com/aurora-is-near/fast-bridge-protocol/pull/113/commits/f6bd540da9b60aefd97f5c38e1155c965bb4affa 

https://github.com/aurora-is-near/fast-bridge-protocol/pull/113/commits/3bcebe661a3a0ccf60e139020d774085a4354940

File: aurora/contracts/src/AuroraErc20FastBridge.sol (hunk 2 above)
Comment (olga24912) [reply aurora/contracts/src/AuroraErc20FastBridge.sol]: fixed: https://github.com/aurora-is-near/fast-bridge-protocol/pull/113/commits/3bcebe661a3a0ccf60e139020d774085a4354940

File: aurora/contracts/src/AuroraErc20FastBridge.sol (hunk 3 above)
Comment (olga24912) [reply aurora/contracts/src/AuroraErc20FastBridge.sol]: fixed: https://github.com/aurora-is-near/fast-bridge-protocol/pull/113/commits/3bcebe661a3a0ccf60e139020d774085a4354940

File: aurora/contracts/src/AuroraErc20FastBridge.sol (hunk 4 above)
Comment (olga24912) [reply aurora/contracts/src/AuroraErc20FastBridge.sol]: https://github.com/aurora-is-near/fast-bridge-protocol/pull/113/commits/d099a07c443e096285172fa366e1f09539d73e41

File: aurora/contracts/src/AuroraErc20FastBridge.sol [hunk 6]
```diff
@@ -43,6 +43,7 @@ contract AuroraErc20FastBridge is Initializable, UUPSUpgradeable, AccessControlU
     NEAR public near;
     string public fastBridgeAccountIdOnNear;
     string public auroraEngineAccountIdOnNear;
+    string public nativeAuroraTokenAccountIdOnNear;
```
Comment (karim-en) [inline aurora/contracts/src/AuroraErc20FastBridge.sol:46]: This corrupt the storage, please move it to the last storage slot.

**Review by karim-en — CHANGES_REQUESTED** [review]

File: aurora/contracts/src/AuroraErc20FastBridge.sol (hunk 6 above)
Comment (olga24912) [reply aurora/contracts/src/AuroraErc20FastBridge.sol]: fixed: https://github.com/aurora-is-near/fast-bridge-protocol/pull/113/commits/7dee54bb8a323b0f1fe63d818bf2e305e5e2607f

**Review by karim-en — APPROVED** [review]

**Pull request merged**